src/main/java/com/google/devtools/build/lib/packages/SelectorList.java

```diff
@@ -110,7 +110,12 @@ static SelectorList concat(Object x, Object y) throws EvalException {
 
   @Override
   public SelectorList binaryOp(TokenKind op, Object that, boolean thisLeft) throws EvalException {
-    if (op == TokenKind.PLUS) {
+    if (getNativeType(that).equals(Dict.class)) {
```

Comment:
You'd need to also make the same change to SelectorValue.binaryOp (and ensure it's tested)

Reply:
Updated it but I'm usure how to test it. As far as a can tell, `SelectorValue` objects never live on their own, but only as elements of `SelectorList`, so there would never be a binary operation involving SelectorValue evaluated from Starlark. I could test the native code, but since `SelectorValue.binaryOp` defers to `SelectorList.binaryOp` after this change, I'm not sure how much value that would bring. @tetromino WDYT?
cc @alandonovan since I see a few comment about merging SelectorValue and SelectorList and this seems relevant

Reply:
I'm no longer on the Blaze team and never got as far as merging these two classes ([three](https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/packages/BuildType.java;l=416;bpv=1;bpt=1?q=BuildType.SelectorList&ss=bazel%2Fbazel) actually). I think your analysis is basically right, though I vaguely recall having seen edge cases in which a SelectorValue escapes from a SelectorList and is seen as a first class Starlark value. In other words, if we merged all three classes, some programs might observe an incompatible change, but I don't think that's a strong reason not to make the change, which would be a significant simplification.

@brandjon may remember more.